teamkeel/keel
# feat: token endpoint and id token authentication

We are to replace the `authenticate()` action and password-based authentication with a set of OAuth endpoints decoupled from the customer's APIs and Keel schema. This means much better support to integrate with 3rd party OAuth providers (such as Google and Facebook), alignment with industry guidelines and standards, and a more secure environment for customer's data and applications.  

This PR kicks off some of this work with the following...

Token Endpoint
===
The token endpoint, located at `/auth/token`, will now handle authentication requests.  It currently accepts the `token-exchange` grant for ID Token authentication (see next section).  The token endpoint will always return the following JSON response:

```
{
    "access_token": "eyJhbGciOiJSUzI...",
    "token_type": "Bearer",
    "expires_in": 86400,
    "refresh_token": "FejvZRD2jBLdoQhvpj"
}
```
**Note:** refresh tokens are not available yet.

ID Token Authentication
===
Users can authenticate with a valid ID token from a registered OpenID Connect provider without needing to configure a client ID or secret with Keel.

This performs all the necessary steps to validate and verify the legitimacy of the ID token.

Access tokens issued from the token endpoint are now decoupled from the original issuer, and they should be used in tandem with refresh tokens.

Testability
===
The majority of my effort here was making this work testable.  I've written a OIDC test server, built on top of @tomfrew's previous work.  This lets us mock a provider for testing purposes.  I think this will become very important as we start to support more flows.  Auth cannot let our users down.

Telemetry
===

The telemetry was quite noisy before.  It's a bit lighter now:
<img width="823" alt="image" src="https://github.com/teamkeel/keel/assets/6212830/fb57f995-0278-43e5-9fed-24713a1c9cf6">


Outstanding
===
 - `aud` verification during ID Token authentication
 - Refresh tokens and refresh token rotation
 - keelconfig.yaml configuration for OIDC providers
 - Add the standard OIDC user claims to `Identity` and update them on authentication
 - Fetching data from the OIDC userinfo endpoint
 - The `createIfNotExists` concept in keelconfig.yaml
 - Caching the OpenID Connect profiles and JWKS public set

Questions to address
===

 - Should it be a minimum requirement that ID tokens provide us with an `email` claim?  
 - Should the issuer of the access tokens be the environment's domain?  Or just `https://keel.so`?
